### PR TITLE
[Framework] Add xml format to export

### PR DESCRIPTION
**Pull request by rim-odoo**
- Task: https://www.odoo.com/web#id=8803&view_type=form&model=project.task
- Pad: https://pad.odoo.com/p/r.GQ9qOYpLqUjA9L7D

### CODE REVIEW

File: openerp/addons/base/res/res_config.py [hunk 1]
```diff
@@ -704,4 +704,4 @@ def get_config_warning(self, cr, msg, context=None):
         # 3/ substitute and return the result
         if (action_id):
             return exceptions.RedirectWarning(msg % values, action_id, _('Go to the configuration panel'))
-        return exceptions.UserError(msg % values)
\ No newline at end of file
+        return exceptions.UserError(msg % values)
```
Comment (rim-odoo) [inline openerp/addons/base/res/res_config.py]: not important: you are adding a diff to a file you did not want to modify

File: addons/web/controllers/main.py [hunk 2]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+
+        if ids:
+            records = request.env[model].browse(ids)
+        else:
+            records = request.env[model].search(domain)
+
+        self.extract_types(map(operator.itemgetter('name'), fields), model)
+
+        self.seen_ids = defaultdict(dict)
+        self.root = etree.Element('odoo')
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: Uh… no, just no. Pass these things explicitly to the export_record method if you need them there.

File: addons/web/controllers/main.py [hunk 3]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+                    if attrs['type'] in ("one2many", "many2many", "many2one"):
+                        models_types[cur_model][tok] = attrs['type']
+                        models_comodels[cur_model][tok] = attrs['relation']
+                    else:
+                        models_types[cur_model][tok] = attrs['type']
+                if tok in models_comodels[cur_model]:
+                    cur_model = models_comodels[cur_model][tok]
+
+        self.models_types = models_types
+        self.models_comodels = models_comodels
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: That's horrifying. Don't go randomly adding new parameters to singleton-ish objects, that's insane.

File: addons/web/controllers/main.py [hunk 4]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+    def filename(self, base):
+        return base + '.xml'
+
+    def extract_types(self, field_names, model):
+        """Extract models data for given fields
+
+        Update object types and comodels corresponding to a list of fields.
+
+        Args:
+            field_names: A list of fields.
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: Wrong format, not sure where you got that one.

File: addons/web/controllers/main.py [hunk 5]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+            return self.seen_ids[model][record.id]
+
+        xml_id = record._BaseModel__export_xml_id()
+        start_no_xml_id = '__export__.'
+        if xml_id.startswith(start_no_xml_id):
+            name = record.name_get()
+            if name:
+                xml_id = xml_id[len(start_no_xml_id):]+'_'+re.sub('[^a-zA-Z0-9-]', '_', name[0][1])
+            else:
+                xml_id = xml_id[len(start_no_xml_id):]
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: Might be time to create "official" internal APIs for querying and generating nice external ids /cc @odony

File: addons/web/controllers/main.py [hunk 6]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+        self.seen_ids[model][record.id] = xml_id
+
+        record_element = etree.Element('record', model=model, id=xml_id)
+
+        types = self.models_types[model]
+        comodels = self.models_comodels[model]
+
+        for field_name in types.keys():
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: why .keys()?

File: addons/web/controllers/main.py [hunk 7]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+            field = etree.SubElement(record_element, 'field', name=field_name)
+
+            if types[field_name] in ("one2many", "many2many", "many2one") and len(getattr(record, field_name)):
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: Doesn't `record[field_name]` work? Also would be nice to call it only once and put the result in a local variable.

File: addons/web/controllers/main.py [hunk 8]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+                if types[field_name] == 'many2one':
+                    field.set('ref', self.export_record(getattr(record, field_name), comodels[field_name]))
+                else:
+                    field.set('eval', '['+', '.join(
+                        '(4,ref(\''+self.export_record(sub_record, comodels[field_name])+'\'))'
+                        for sub_record in getattr(record, field_name)
+                    )+']')
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: 1. string formatting is often clearer than concatenation
2. if you need single quotes inside a string, you can use double quotes for the string itself (and the other way around)

File: addons/web/controllers/main.py [hunk 9]
```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
...
+            else:
+                field.text = unicode(getattr(record, field_name))
```
Comment (xmo-odoo) [inline addons/web/controllers/main.py]: I think this is going to fail for a bunch of field types e.g. boolean.

**Comment by nle-odoo** [comment]
thanks for all these valid points @xmo-odoo and @rim-odoo , I will see to it on tuesday.

**Comment by nle-odoo** [comment]
I did most changes, i will try and test further for different types of field.

**Pull request closed without merging**